org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/rename/RenameService2.xtend

```diff
@@ -77,31 +114,137 @@ class RenameService2 implements IRenameService2 {
 			} else {
 				issueAcceptor.add(FATAL, 'Loaded resource is not an XtextResource', context.resource.URI)
 			}
+			issueAcceptor.checkSeverity
 			return workspaceEdit
+		].exceptionally [ exception |
+			val rootCause = Throwables.getRootCause(exception);
+			if (rootCause instanceof FileNotFoundException) {
+				if (shouldPrepareRename) {
+					return null
+				}
+			}
+			throw exception
 		].get
 	}
 
 	protected def EObject getElementAtOffset(XtextResource xtextResource, Document document, Position caretPosition) {
 		val caretOffset = document.getOffSet(caretPosition)
-		return xtextResource.getElementWithIdentifierAt(caretOffset) 
+		return xtextResource.getElementWithIdentifierAt(caretOffset)
 			?: xtextResource.getElementWithIdentifierAt(caretOffset - 1)
 	}
-	
+
 	protected def getElementWithIdentifierAt(XtextResource xtextResource, int offset) {
 		if (offset >= 0) {
 			val rootNode = xtextResource?.parseResult?.rootNode
 			if (rootNode !== null) {
 				val leaf = NodeModelUtils.findLeafNodeAtOffset(rootNode, offset)
 				if (leaf !== null && leaf.isIdentifier) {
-					return xtextResource.resolveElementAt(offset)	
+					return xtextResource.resolveElementAt(offset)
 				}
 			}
 		}
 		return null
 	}
-	
+
 	protected def isIdentifier(ILeafNode leafNode) {
 		return (leafNode.grammarElement instanceof RuleCall || leafNode.grammarElement instanceof CrossReference)
-			&& !tokenUtil.isWhitespaceOrCommentNode(leafNode) 
+			&& !tokenUtil.isWhitespaceOrCommentNode(leafNode)
+	}
+
+	override prepareRename(PrepareRenameOptions options) {
+		val uri = options.params.textDocument.uri
+		val shouldPrepareRename = options.languageServerAccess.shouldPrepareRename
```

Comment:
If that is false, shouldn't we skip all the rest of this method?

Reply:
No, you can invoke `textDocument/prepareRename` without calling `textDocument/rename`.

Reply:
Yes, you're right, but if `shouldPrepareRename` is `false`, we still have to throw an exception if the resource does not exist. It is not clear from the spec, whether calling `textDocument/prepareRename` itself is a valid call or it is always an implicit precheck before `textDocument/rename`.

Reply:
It is [done](https://github.com/eclipse/xtext-core/commit/6bb48ebd57781ecb175934e3bab73ab1745d87e8#diff-f0264dd6c439d644cd998f039eba6d21R160).